macropy/imacro.py

```diff
@@ -0,0 +1,116 @@
+# -*- coding: utf-8 -*-
+"""IPython extension for a MacroPy-enabled REPL.
+
+To enable::
+
+    %load_ext macropy.imacro
+
+To autoload it at IPython startup, put this into your ``ipython_config.py``::
+
+    c.InteractiveShellApp.extensions = ["macropy.imacro"]
+"""
+
+import ast
+import importlib
+from collections import OrderedDict
+from functools import partial
+
+from IPython.core.error import InputRejected
+
+from macropy import __version__ as macropy_version
+from macropy.core.macros import ModuleExpansionContext, detect_macros
+
+_placeholder = "<interactive input>"
+_instance = None
+
+def load_ipython_extension(ipython):
+    print("MacroPy {} -- Syntactic macros for Python.".format(macropy_version))
+    global _instance
+    if not _instance:
+        _instance = IMacroPyExtension(shell=ipython)
+
+def unload_ipython_extension(ipython):
+    global _instance
+    _instance = None
+
+class MacroTransformer(ast.NodeTransformer):
+    def __init__(self, extension_instance, *args, **kwargs):
+        super().__init__(*args, **kwargs)
+        self.ext = extension_instance
+        self.bindings = OrderedDict()
+
+    def visit(self, tree):
+        try:
+            bindings = detect_macros(tree, '__main__', reload=True)  # macro imports
+            if bindings:
+                self.ext.macro_bindings_changed = True
+                for fullname, macro_bindings in bindings:
+                    mod = importlib.import_module(fullname)
+                    self.bindings[fullname] = (mod, macro_bindings)
+            newtree = ModuleExpansionContext(tree, self.ext.src, self.bindings.values()).expand_macros()
+            self.ext.src = _placeholder
+            return newtree
+        except Exception as err:
+            # see IPython.core.interactiveshell.InteractiveShell.transform_ast()
+            raise InputRejected(*err.args)
+
+class IMacroPyExtension:
+    def __init__(self, shell):
+        self.src = _placeholder
+        self.shell = shell
+
+        self.shell.get_ipython().events.register('pre_run_cell', self._get_source_code)
+#        # TODO: maybe use something like the following instead, to get the
+#        # source code after any string-based transformers have run:
+#        if hasattr(self.shell, "input_transformers_post"):  # IPython 7.0+ with Python 3.5+
```

Comment:
why not make it conditional? then newer versions get the newer code, like maybe

```python
if sys.version_info[0:2] >= (3, 5) and \
    tuple(map(lambda x: int(x), IPython.__version__.split("."))) >= (7, 0):
    if hasattr(self.shell, "input_transformers_post"):  # IPython 7.0+ with Python 3.5+
        def get_source_code(lines):
            self.src = lines
            return lines
        self.shell.input_transformers_post.append(get_source_code)
else:
    self.shell.get_ipython().events.register('pre_run_cell', self._get_source_code)
```

Reply:
(not sure how to get the IPython API version but that works in a notebook https://github.com/ipython/ipython/issues/3025)

Reply:
Mm, why not.

I was originally going for simplicity, but then it hit me that strictly speaking, `pre_run_cell` gets the source text too early, hence the commented-out copy'n'paste (from the first version, before this was changed into an IPython plugin).

For most uses the difference doesn't matter, but the best source text to use for error reporting is that just before AST transformers start running - after all string transformers have run. The commented-out solution does that in IPython 7.0 and later. (The API for string transformers [completely changed in 7.0](https://ipython.readthedocs.io/en/stable/config/inputtransforms.html); I think there's no point in going out of our way to support old versions.)

The `hasattr` actually already duck-tests for the relevant feature of the new API.

Reply:
Well, https://jupyter.org/try uses 6.5.0, so it seems reasonable to me that a lot of users are on 6.x. But, there's probably nothing wrong with just requiring IPython 7 for MacroPy and throwing like a `IPython.core.error.IPythonCoreError: IPython not >= 7.0` or something

Reply:
I think the conditional approach is fine - what I meant is, since in most usage the difference between the original raw source text and the string-transformed source text does not matter, it should be fine in practice to use `pre_run_cell` for versions below 7.0, and take advantage of the new API for 7.0 and later.

I have tested this in 6.5, and didn't notice any problems with the `pre_run_cell` solution. Theoretically, there could be a problem if the same input cell contains e.g. IPython magics (that might insert more lines when they expand) and MacroPy macros. The solution using the 7.0 API gets this corner case right.

I haven't looked at all the details, but *"Any third party code extending input transformation will need to be rewritten"* [[source]](https://ipython.readthedocs.io/en/stable/config/inputtransforms.html#string-based-transformations) sounds like more trouble than it's worth to support a corner case in an old version. :)

I'll push a new version shortly. Thanks for reviewing and for noticing this!